zerver/lib/test_classes.py

```diff
@@ -1384,6 +1394,55 @@ def setUp(self) -> None:
         super().setUp()
         self.url = self.build_webhook_url()
 
+        if self.WEBHOOK_DIR_NAME is not None and self.VIEW_FUNCTION_NAME is not None:
+            # The webhook_view decorator adds all_event_types that is passed as an argument
+            # (or None by default) to the decorated view function as an attribute.
+            # If VIEW_FUNCTION_NAME is explicitly specified and WEBHOOK_DIR_NAME is not None,
+            # an exception will be raised when a test triggers events that are not explicitly
+            # specified in the webhook_view decorator.
+            function = import_string(
+                f"zerver.webhooks.{self.WEBHOOK_DIR_NAME}.view.{self.VIEW_FUNCTION_NAME}"
+            )
+            all_event_types = None
+
+            if hasattr(function, "_all_event_types"):
+                all_event_types = function._all_event_types
+
+            if all_event_types is None:
+                return
+
+            def side_effect(*args: Any, **kwargs: Any) -> None:
+                complete_event_type = (
+                    kwargs.get("complete_event_type")
+                    if len(args) < 5
+                    else args[4]  # complete_event_type is the argument at index 4
+                )
+                if (
+                    complete_event_type is not None
+                    and all_event_types is not None
+                    and complete_event_type not in all_event_types
+                ):
+                    raise Exception(
+                        f"""
+Your test code triggered the event "{complete_event_type}" which was not registered
+by using webhook_view function decorator.
+
+It is likely that you used the incoming webhook event filtering system without adding this
+event to all_event_types in the webhook_view function decorator, or you incorrectly
+parsed the event type when calling check_send_webhook_message.
```

Comment:
I think this is confusing, in that the developer may not have used the incoming webhook event filtering system, if they're adding new events to an existing webhook.

Proposed new language:
```
Error: This test triggered a message using the event "{complete_event_type}", which was not properly
registered via the @webhook_view(..., event_types=[...]). These registrations are important for Zulip
self-documenting the supported event types for this integration.

You can fix this by adding "{complete_event_type}" to ALL_EVENT_TYPES for this webhook. 
See {documentation_url} for details.
```

(We can leave off the last line for now, and add it when we add documentation for this infrastructure).

And then once we've added the docs and event_types declarations for most webhooks, we can have a separate exception for "webhook not on an exclude list of webhooks that are known to not declare event types", to prevent new webhooks from being introduced without this feature.

Reply:
OK, this sounds reasonable.